compiler/toc_hir_lowering/src/scopes.rs

```diff
@@ -27,6 +27,14 @@ pub(crate) enum ScopeKind {
     Module,
     Block,
     Loop,
+    Subprogram,
+    SubprogramHeader,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum LookupKind {
+    Normal,
+    RedeclChecking,
```

Comment:
Undecided on name.
There's two main types of lookup that we can do:

- Lookup on use
- Lookup on definition

So naming after that might be ideal.